File: reference/go/aep/delivery_journal.go
package aep

import (
	"strconv"
	"strings"
	"time"
)

type DeliveryJournal struct {
	streamId string
	events   []map[string]any
	sequence int
}

func NewDeliveryJournal(streamId string) *DeliveryJournal {
	return &DeliveryJournal{
		streamId: streamId,
	}
}

func (j *DeliveryJournal) NextSequence() int {
	j.sequence++
	return j.sequence
}

func (j *DeliveryJournal) Append(event map[string]any) int {
	seq := j.NextSequence()
	record := make(map[string]any)
	for k, v := range event {
		record[k] = v
	}
	record["_journal_sequence"] = seq
	record["_journal_cursor"] = j.streamId + ":" + strconv.Itoa(seq)
	record["_journal_appendedAt"] = time.Now().UTC().Format(time.RFC3339)
	j.events = append(j.events, record)
	return seq
}

func (j *DeliveryJournal) Replay(cursor string) []map[string]any {
	if cursor == "" {
		result := make([]map[string]any, len(j.events))
		copy(result, j.events)
		return result
	}
	parts := strings.Split(cursor, ":")
	sinceSeq := 0
	if len(parts) > 1 {
		sinceSeq, _ = strconv.Atoi(parts[1])
	}
	var result []map[string]any
	for _, e := range j.events {
		if e["_journal_sequence"].(int) > sinceSeq {
			result = append(result, e)
		}
	}
	return result
}

func (j *DeliveryJournal) ReplaySinceSequence(seq int) []map[string]any {
	var result []map[string]any
	for _, e := range j.events {
		if e["_journal_sequence"].(int) > seq {
			result = append(result, e)
		}
	}
	return result
}

func (j *DeliveryJournal) Purge(cursor string) int {
	parts := strings.Split(cursor, ":")
	beforeSeq := 0
	if len(parts) > 1 {
		beforeSeq, _ = strconv.Atoi(parts[1])
	}
	removed := 0
	for len(j.events) > 0 && j.events[0]["_journal_sequence"].(int) <= beforeSeq {
		j.events = j.events[1:]
		removed++
	}
	return removed
}

func (j *DeliveryJournal) GetStats() map[string]any {
	stats := make(map[string]any)
	stats["totalEvents"] = len(j.events)
	if len(j.events) > 0 {
		stats["oldestSequence"] = j.events[0]["_journal_sequence"]
		stats["newestSequence"] = j.events[len(j.events)-1]["_journal_sequence"]
	} else {
		stats["oldestSequence"] = nil
		stats["newestSequence"] = nil
	}
	return stats
}

File: implementations/python/src/aep/delivery_journal.py
from datetime import datetime, timezone


def _now() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


class DeliveryJournal:
    def __init__(self, stream_id: str = "stream_01"):
        self._stream_id = stream_id
        self._events: list[dict] = []
        self._sequence = 0

    def next_sequence(self) -> int:
        self._sequence += 1
        return self._sequence

    def append(self, event: dict) -> int:
        seq = self.next_sequence()
        record = {
            **event,
            "_journal_sequence": seq,
            "_journal_cursor": f"{self._stream_id}:{seq}",
            "_journal_appendedAt": _now(),
        }
        self._events.append(record)
        return seq

    def replay(self, cursor: str | None = None) -> list[dict]:
        if cursor is None:
            return list(self._events)
        parts = cursor.split(":")
        since_seq = int(parts[1]) if len(parts) > 1 else 0
        return [e for e in self._events if e["_journal_sequence"] > since_seq]

    def replay_since_sequence(self, seq: int) -> list[dict]:
        return [e for e in self._events if e["_journal_sequence"] > seq]

    def purge(self, cursor: str) -> int:
        parts = cursor.split(":")
        before_seq = int(parts[1]) if len(parts) > 1 else 0
        removed = 0
        while self._events and self._events[0]["_journal_sequence"] <= before_seq:
            self._events.pop(0)
            removed += 1
        return removed

    def get_stats(self) -> dict:
        return {
            "totalEvents": len(self._events),
            "oldestSequence": self._events[0]["_journal_sequence"] if self._events else None,
            "newestSequence": self._events[-1]["_journal_sequence"] if self._events else None,
        }
